x/ccv/provider/keeper/grpc_query_test.go

```diff
@@ -107,32 +109,89 @@ func TestQueryConsumerValidators(t *testing.T) {
+	pk.SetConsumerPowerShapingParameters(ctx, consumerId, types.PowerShapingParameters{Top_N: 0})
 
-	providerAddr2 := types.NewProviderConsAddress([]byte("providerAddr2"))
-	consumerKey2 := cryptotestutil.NewCryptoIdentityFromIntSeed(2).TMProtoCryptoPublicKey()
-	consumerValidator2 := types.ConsensusValidator{ProviderConsAddr: providerAddr2.ToSdkConsAddr(), Power: 2, PublicKey: &consumerKey2}
+	// create bonded validators
+	val1 := createStakingValidator(ctx, mocks, 1, 1, 1)
+	pk1, _ := val1.CmtConsPublicKey()
+	valConsAddr1, _ := val1.GetConsAddr()
+	providerAddr1 := types.NewProviderConsAddress(valConsAddr1)
+
+	val2 := createStakingValidator(ctx, mocks, 1, 2, 2)
+	pk2, _ := val2.CmtConsPublicKey()
+	valConsAddr2, _ := val2.GetConsAddr()
+	providerAddr2 := types.NewProviderConsAddress(valConsAddr2)
 
-	expectedResponse := types.QueryConsumerValidatorsResponse{
+	// set expectation to return bonded validators
+	testkeeper.SetupMocksForLastBondedValidatorsExpectation(mocks.MockStakingKeeper, 2, []stakingtypes.Validator{val1, val2}, []int64{1, 2}, -1) // -1 to allow the calls "AnyTimes"
+
+	// set max provider consensus vals to include all validators
+	params := pk.GetParams(ctx)
+	params.MaxProviderConsensusValidators = 2
+	pk.SetParams(ctx, params)
+
+	// expect no validator to be returned since the consumer is Opt-In
+	res, err := pk.QueryConsumerValidators(ctx, &req)
+	require.NoError(t, err)
+	require.Len(t, res.Validators, 0)
+
+	// opt in one validator
+	pk.SetOptedIn(ctx, consumerId, providerAddr1)
+
+	// expect opted-in validators
+	res, err = pk.QueryConsumerValidators(ctx, &req)
+	require.NoError(t, err)
+	require.Len(t, res.Validators, 1)
+	require.Equal(t, res.Validators[0].ProviderAddress, providerAddr1.String())
+
+	// update consumer TopN param
+	pk.SetConsumerPowerShapingParameters(ctx, consumerId, types.PowerShapingParameters{Top_N: 50})
+
+	// expect both opted-in and topN validator
+	expRes := types.QueryConsumerValidatorsResponse{
 		Validators: []*types.QueryConsumerValidatorsValidator{
-			{providerAddr1.String(), &consumerKey1, 1},
-			{providerAddr2.String(), &consumerKey2, 2},
+			{ProviderAddress: providerAddr1.String(), ConsumerKey: &pk1, Power: 1},
+			{ProviderAddress: providerAddr2.String(), ConsumerKey: &pk2, Power: 2},
 		},
 	}
+	res, err = pk.QueryConsumerValidators(ctx, &req)
+	require.NoError(t, err)
+	require.Equal(t, &expRes, res)
 
-	// set up the client id so the chain looks like it "started"
-	pk.SetConsumerClientId(ctx, consumerId, "clientID")
-	pk.SetConsumerValSet(ctx, consumerId, []types.ConsensusValidator{consumerValidator1, consumerValidator2})
+	// advance time so that the chain looks like it "launched"
+	ctx = ctx.WithBlockTime(ctx.BlockTime().Add(time.Hour))
 
-	res, err := pk.QueryConsumerValidators(ctx, &req)
+	// expect an empty consumer valset
+	// since neither QueueVSCPackets or MakeConsumerGenesis was called at this point
+	res, err = pk.QueryConsumerValidators(ctx, &req)
```

Comment:
Does this mean that from `spawnTime` till the actual launch, querying consumer validators would return an empty set? Is this what we want? @MSalopek

Reply:
No. Here we have a special case where the spawn time of the consumer chain has elapsed but has never gone through `MakeConsumerGenesis` - which should never happen in production.

Reply:
With permissionless we could have 10k consumer chains wanting to spawn and because we launch the chain in the `BeginBlock` we probably intend to this in batches, e.g., 200 chains launch in block 1, the next 200 chains launch in block 2, etc. So, I guess the `spawnTime` might have passed but we have not executed `MakeConsumerGenesis` yet. Does that sound reasonable?

Reply:
Updated the logic to check the consumer phase instead of the spawn time to avoid this issue.